### ISSUE TITLE
Prediction for way<5

### ISSUE THREAD

**Issue by manashb96** [cfifty/CAML#15]
Firstly awesome project and great work guys. 

So lets say I try making  predictions for way<5. The model is able to make the predictions, but the transformer encoder output is of shape `torch.Size([1, 5])`. If i try changing the num_classes in the linear layer of the transformer model it doesn't work as the checkpoint has a 5 way training process.
`        size mismatch for transformer_encoder.output_proj.weight: copying a param with shape torch.Size([5, 1024]) from checkpoint, the shape in current model is torch.Size([2, 1024]).`

Hence my question is assuming we do a n shot inference of the model where n<5. Will it perform as per expectation even though the encoder model is giving a 5 way output or should we go for retraining of the transformer model, or is there a different method of tackling the issue, where we want the logits to be of proper shape and not fixed at 5 dim.

For better understanding, lets say I am doing 10 shot 2 way classification between apples (0th support class) and oranges (1st support class). In most cases the argmax  of the logits is limited to the 0th and 1st index of the logit variable, but in certain use cases the argmax may come in either the 3rd, 4th or 5th index as the logit output is of shape (1,5).

**Comment by cfifty** [comment]
Hey! You want to load the model as normal, just ignore the predictions for the 3-5 positions. The input to the in-context learning sequence model will look something like: 

[(example_1, label_1), ..., (example_5, label_1), (example_6, label_2), ..., (example_10, label_2)]

The output will be of shape (b, 5) where b is your batch size. Just ignore the last 3 dimensions. If you want to be a bit more accurate, directly drop the values at these indices before taking the softmax:

```
logits = model(x)  # Get the model's predictions: will be (b,5)
logits = logits[:,:2]  # Drop the final 3 predictions: will be (b,2)
preds = torch.nn.functional.softmax(logits, dim=-1)  # Take the softmax of the revised predictions.
```

**Comment by yaochx** [comment]
I have a question: 
when input way = 2,  sometimes the sequential model output are > 2; the internal logits are like this:
```
[ 0.4323, -2.6092, -0.3186,  1.1073,  0.5364],
[-2.6239, -1.2488,  0.9519,  0.9190,  0.9721]
```
the last 3 dimensions' numbers  don't seem random; they appear to follow a calculated logic from output_prjection. Why can we ignore the last 3 dimension output from linear projection's, can you give an explaination kindly? @cfifty 

Many thanks for you and your awesome work.

**Comment by cfifty** [comment]
Hey! You can view the output projection as partitioning the ambient space (768-dimensions) with 5 hyperplanes. The last linear layer defines the normal vector to those 5 hyperplanes (so the columns of the 768x5 linear projection). 

The forward pass of CAML pulls the query (i.e. the unlabeled example) towards one of the classes in the support set (i.e. the labeled examples) based on the features that compose each support set class. Since you do not have any examples from support sets numbered 3-5, there should be no "pulling" towards these classes, and the logits for them should be ignored. 

You will still have the comparison between the internal representation and the normal vector for these hyperplanes; however, because the model was always trained in a 5-shot scenario.

**Comment by yaochx** [comment]
Yes, almost understood.

In my case, when the input way is set to 2, there’s approximately a 5% chance that I should disregard the last 3 logits.

Is this a good idea or pointless to fine-tune the MLP to adapt to way variation?

**Comment by cfifty** [comment]
Can you expand on:
```
In my case, when the input way is set to 2, there’s approximately a 5% chance that I should disregard the last 3 logits.
```
I do not quite understand what you are asking. Perhaps provide a simple example?

**Comment by yaochx** [comment]
During training, can we randomly sample various configurations (e.g., 2-way, 3-way, 5-way) and train the MLP to accommodate a dynamic number of logits? I believe specific label context information can be integrated into the transformer encoder output (query's high-dimensional features) embedding space.

**Comment by cfifty** [comment]
Hey! Yes; you can do this during training of the Transformer Encoder. My prior is this will be very similar to simply training the full 5-way, but many great research findings start with unexpected results. It's possible the transformer encoder / sequence model overfits to the 5-way context and would be unable to generalize to other configurations.

**Comment by yaochx** [comment]
Okay, thanks for your attention. We can close this issue again.
